File: packages/boss-crawler-worker/src/protocol.ts
import { z } from "zod";

export const LoginStartPayloadSchema = z.object({
  executable_path: z.string().optional(),
  user_data_dir: z.string().optional(),
});

export const SessionStatePayloadSchema = z.object({
  cookies: z.any(),
  local_storage: z.any(),
});

export const SearchTaskPayloadSchema = z.object({
  keywords: z.array(z.string().min(1)).min(1),
  filters: z.any().optional().default({}),
  limits: z.any().optional().default({}),
  mode: z.string().optional(),
});

export const CrawlManualStartPayloadSchema = z.object({
  session: SessionStatePayloadSchema,
});

export const CrawlAutoStartPayloadSchema = z.object({
  session: SessionStatePayloadSchema,
  task: SearchTaskPayloadSchema,
});

export const BossMetaSyncPayloadSchema = z.object({
  session: SessionStatePayloadSchema,
});

export const AiAnalyzePayloadSchema = z.object({
  resume_text: z.string().min(1),
  context_text: z.string().optional(),
  resume_files: z.string().optional(),
  job_detail: z.any(),
});

export const AiAnalyzeGroupPayloadSchema = z.object({
  context_text: z.string().min(1),
  jobs: z.array(z.any()).min(1),
});

export const ResumeDiagnosePayloadSchema = z.object({
  resume_text: z.string().min(1),
  context_text: z.string().optional(),
  resume_files: z.string().optional(),
});

export const ResumeRewriteModulePayloadSchema = z.object({
  module: z.enum(["summary", "projects", "experience", "skills"]),
  resume_text: z.string().min(1),
  context_text: z.string().optional(),
  resume_files: z.string().optional(),
  module_input: z.string().min(1),
  confirmed_summary: z.string().optional(),
  confirmed_projects: z.string().optional(),
  confirmed_experience: z.string().optional(),
  confirmed_skills: z.string().optional(),
});

export const CommandInSchema = z.union([
  z.object({ type: z.literal("LOGIN_START"), payload: LoginStartPayloadSchema }),
  z.object({
    type: z.literal("CRAWL_MANUAL_START"),
    payload: CrawlManualStartPayloadSchema,
  }),
  z.object({
    type: z.literal("CRAWL_AUTO_START"),
    payload: CrawlAutoStartPayloadSchema,
  }),
  z.object({
    type: z.literal("BOSS_META_SYNC"),
    payload: BossMetaSyncPayloadSchema,
  }),
  z.object({ type: z.literal("AI_ANALYZE"), payload: AiAnalyzePayloadSchema }),
  z.object({ type: z.literal("AI_ANALYZE_GROUP"), payload: AiAnalyzeGroupPayloadSchema }),
  z.object({ type: z.literal("RESUME_DIAGNOSE"), payload: ResumeDiagnosePayloadSchema }),
  z.object({ type: z.literal("RESUME_REWRITE_MODULE"), payload: ResumeRewriteModulePayloadSchema }),
  z.object({ type: z.literal("STOP") }),
  z.object({ type: z.literal("PAUSE") }),
  z.object({ type: z.literal("RESUME") }),
]);

export type CommandIn = z.infer<typeof CommandInSchema>;

export const LogPayloadSchema = z.object({
  level: z.string(),
  message: z.string(),
  ts: z.string().optional(),
});

export const ProgressPayloadSchema = z.object({
  keyword: z.string().optional(),
  current_page: z.number().int().nonnegative().optional(),
  captured_job_list: z.number().int().nonnegative().optional(),
  captured_job_detail: z.number().int().nonnegative().optional(),
});

export const LoginStatusPayloadSchema = z.object({
  status: z.string(),
  message: z.string().optional(),
});

export const CookieCollectedPayloadSchema = z.object({
  cookies: z.any(),
  local_storage: z.any(),
});

export const JobListCapturedPayloadSchema = z.object({
  keyword: z.string().optional(),
  filters: z.any().optional(),
  raw: z.any(),
});

export const JobDetailCapturedPayloadSchema = z.object({
  encrypt_job_id: z.string().min(1),
  zp_data: z.any(),
});

export const AiResultPayloadSchema = z.object({
  result: z.any(),
});

export const BossMetaSyncedPayloadSchema = z.object({
  synced_at: z.string().optional(),
  city_group: z.any().optional(),
  filter_conditions: z.any().optional(),
  industry_filter_exemption: z.any().optional(),
});

export const ErrorPayloadSchema = z.object({
  message: z.string(),
  stack: z.string().optional(),
});

export const EventOutSchema = z.union([
  z.object({ type: z.literal("LOG"), payload: LogPayloadSchema }),
  z.object({ type: z.literal("PROGRESS"), payload: ProgressPayloadSchema }),
  z.object({ type: z.literal("LOGIN_STATUS"), payload: LoginStatusPayloadSchema }),
  z.object({ type: z.literal("COOKIE_COLLECTED"), payload: CookieCollectedPayloadSchema }),
  z.object({ type: z.literal("JOB_LIST_CAPTURED"), payload: JobListCapturedPayloadSchema }),
  z.object({ type: z.literal("JOB_DETAIL_CAPTURED"), payload: JobDetailCapturedPayloadSchema }),
  z.object({ type: z.literal("AI_RESULT"), payload: AiResultPayloadSchema }),
  z.object({ type: z.literal("BOSS_META_SYNCED"), payload: BossMetaSyncedPayloadSchema }),
  z.object({ type: z.literal("FINISHED") }),
  z.object({ type: z.literal("ERROR"), payload: ErrorPayloadSchema }),
]);

export type EventOut = z.infer<typeof EventOutSchema>;

File: src-tauri/src/ipc/protocol.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LoginStartPayload {
  #[serde(default)]
  #[serde(skip_serializing_if = "Option::is_none")]
  pub executable_path: Option<String>,
  #[serde(default)]
  #[serde(skip_serializing_if = "Option::is_none")]
  pub user_data_dir: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionStatePayload {
  pub cookies: Value,
  pub local_storage: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CrawlAutoStartPayload {
  pub session: SessionStatePayload,
  pub task: SearchTaskPayload,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CrawlManualStartPayload {
  pub session: SessionStatePayload,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BossMetaSyncPayload {
  pub session: SessionStatePayload,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchTaskPayload {
  pub keywords: Vec<String>,
  #[serde(default)]
  pub filters: Value,
  #[serde(default)]
  pub limits: Value,
  #[serde(default)]
  #[serde(skip_serializing_if = "Option::is_none")]
  pub mode: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AiAnalyzePayload {
  pub resume_text: String,
  #[serde(default)]
  #[serde(skip_serializing_if = "Option::is_none")]
  pub context_text: Option<String>,
  #[serde(default)]
  #[serde(skip_serializing_if = "Option::is_none")]
  pub resume_files: Option<String>,
  pub job_detail: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AiAnalyzeGroupPayload {
  pub context_text: String,
  pub jobs: Vec<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResumeDiagnosePayload {
  pub resume_text: String,
  #[serde(default)]
  #[serde(skip_serializing_if = "Option::is_none")]
  pub context_text: Option<String>,
  #[serde(default)]
  #[serde(skip_serializing_if = "Option::is_none")]
  pub resume_files: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResumeRewriteModulePayload {
  pub module: String,
  pub resume_text: String,
  #[serde(default)]
  #[serde(skip_serializing_if = "Option::is_none")]
  pub context_text: Option<String>,
  #[serde(default)]
  #[serde(skip_serializing_if = "Option::is_none")]
  pub resume_files: Option<String>,
  pub module_input: String,
  #[serde(default)]
  #[serde(skip_serializing_if = "Option::is_none")]
  pub confirmed_summary: Option<String>,
  #[serde(default)]
  #[serde(skip_serializing_if = "Option::is_none")]
  pub confirmed_projects: Option<String>,
  #[serde(default)]
  #[serde(skip_serializing_if = "Option::is_none")]
  pub confirmed_experience: Option<String>,
  #[serde(default)]
  #[serde(skip_serializing_if = "Option::is_none")]
  pub confirmed_skills: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "payload")]
pub enum CommandIn {
  #[serde(rename = "LOGIN_START")]
  LoginStart(LoginStartPayload),
  #[serde(rename = "CRAWL_MANUAL_START")]
  CrawlManualStart(CrawlManualStartPayload),
  #[serde(rename = "CRAWL_AUTO_START")]
  CrawlAutoStart(CrawlAutoStartPayload),
  #[serde(rename = "BOSS_META_SYNC")]
  BossMetaSync(BossMetaSyncPayload),
  #[serde(rename = "AI_ANALYZE")]
  AiAnalyze(AiAnalyzePayload),
  #[serde(rename = "AI_ANALYZE_GROUP")]
  AiAnalyzeGroup(AiAnalyzeGroupPayload),
  #[serde(rename = "RESUME_DIAGNOSE")]
  ResumeDiagnose(ResumeDiagnosePayload),
  #[serde(rename = "RESUME_REWRITE_MODULE")]
  ResumeRewriteModule(ResumeRewriteModulePayload),
  #[serde(rename = "STOP")]
  Stop,
  #[serde(rename = "PAUSE")]
  Pause,
  #[serde(rename = "RESUME")]
  Resume,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LogPayload {
  pub level: String,
  pub message: String,
  #[serde(default)]
  pub ts: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProgressPayload {
  #[serde(default)]
  pub keyword: Option<String>,
  #[serde(default)]
  pub current_page: Option<u32>,
  #[serde(default)]
  pub captured_job_list: Option<u32>,
  #[serde(default)]
  pub captured_job_detail: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LoginStatusPayload {
  pub status: String,
  #[serde(default)]
  pub message: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CookieCollectedPayload {
  pub cookies: Value,
  pub local_storage: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JobListCapturedPayload {
  #[serde(default)]
  pub keyword: Option<String>,
  #[serde(default)]
  pub filters: Option<Value>,
  pub raw: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JobDetailCapturedPayload {
  pub encrypt_job_id: String,
  pub zp_data: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AiResultPayload {
  pub result: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BossMetaSyncedPayload {
  #[serde(default)]
  #[serde(skip_serializing_if = "Option::is_none")]
  pub synced_at: Option<String>,
  #[serde(default)]
  #[serde(skip_serializing_if = "Option::is_none")]
  pub city_group: Option<Value>,
  #[serde(default)]
  #[serde(skip_serializing_if = "Option::is_none")]
  pub filter_conditions: Option<Value>,
  #[serde(default)]
  #[serde(skip_serializing_if = "Option::is_none")]
  pub industry_filter_exemption: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ErrorPayload {
  pub message: String,
  #[serde(default)]
  pub stack: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "payload")]
pub enum EventOut {
  #[serde(rename = "LOG")]
  Log(LogPayload),
  #[serde(rename = "PROGRESS")]
  Progress(ProgressPayload),
  #[serde(rename = "LOGIN_STATUS")]
  LoginStatus(LoginStatusPayload),
  #[serde(rename = "COOKIE_COLLECTED")]
  CookieCollected(CookieCollectedPayload),
  #[serde(rename = "JOB_LIST_CAPTURED")]
  JobListCaptured(JobListCapturedPayload),
  #[serde(rename = "JOB_DETAIL_CAPTURED")]
  JobDetailCaptured(JobDetailCapturedPayload),
  #[serde(rename = "AI_RESULT")]
  AiResult(AiResultPayload),
  #[serde(rename = "BOSS_META_SYNCED")]
  BossMetaSynced(BossMetaSyncedPayload),
  #[serde(rename = "FINISHED")]
  Finished,
  #[serde(rename = "ERROR")]
  Error(ErrorPayload),
}
